command.go

```diff
@@ -553,6 +590,78 @@ func (inv *Invocation) with(fn func(*Invocation)) *Invocation {
 	return &i2
 }
 
+func (inv *Invocation) doCompletions() []string {
+	prev, cur := inv.curWords()
+	inv.CurWord = cur
+	// If the current word is a flag set using `=`, use it's handler
+	if strings.HasPrefix(cur, "--") && strings.Contains(cur, "=") {
+		if out := inv.equalsFlagCompletions(cur); out != nil {
+			return out
+		}
+	}
+	// If the previous word is a flag, then we're writing it's value
+	// and we should check it's handler
+	if strings.HasPrefix(prev, "--") {
+		if out := inv.flagCompletions(prev); out != nil {
+			return out
+		}
+	}
+	// If the current word is the command, auto-complete it so the shell moves the cursor
+	if inv.Command.Name() == inv.CurWord {
+		return []string{inv.Command.Name()}
+	}
+	var completions []string
+
+	if inv.Command.CompletionHandler != nil {
+		completions = append(completions, inv.Command.CompletionHandler(inv)...)
+	}
+
+	completions = append(completions, DefaultCompletionHandler(inv)...)
+
+	return completions
+}
+
+func (inv *Invocation) flagCompletions(word string) []string {
+	return inv.doFlagCompletions("", word)
+}
+
+func (inv *Invocation) equalsFlagCompletions(word string) []string {
+	words := strings.Split(word, "=")
+	word = words[0]
+	if len(words) > 1 {
+		inv.CurWord = words[1]
+	} else {
+		inv.CurWord = ""
+	}
+	prefix := word + "="
+	return inv.doFlagCompletions(prefix, word)
+}
+
+func (inv *Invocation) doFlagCompletions(prefix, word string) []string {
+	opt := inv.Command.Options.ByFlag(word[2:])
+	if opt == nil {
+		return nil
+	}
+	if opt.CompletionHandler != nil {
+		completions := opt.CompletionHandler(inv)
+		out := make([]string, 0, len(completions))
+		for _, completion := range completions {
+			out = append(out, fmt.Sprintf("%s%s", prefix, completion))
```

Comment:
If we're just returning the completions with the prefix prepended, why even pass prefix down here? We can "factor it out" from both sides of the signature and take away one more concept from this function.

If we handled all of the nuance of using `--flag=abcd` vs `--flag abcd` in one little place on the call stack, the code would be far more readable.